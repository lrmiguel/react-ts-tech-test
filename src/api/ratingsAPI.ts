type RatingsMetadataType = {
  dataSource: string;
  extractDate: string;
  itemCount: number;
  returncode: string;
  totalCount: number;
  totalPages: number;
  pageSize: number;
  pageNumber: number;
};

type LinkType = {
  rel: string;
  href: string;
};

export type EstablishmentsType = {
  establishments: {}[];
  meta: RatingsMetadataType;
  links: LinkType[];
};

export type AuthorityType = {
  Name: string;
  LocalAuthorityId: number;
}

export type AuthoritiesType = {
  authorities: AuthorityType[];
  meta?: RatingsMetadataType;
  links?: LinkType[];
};

export function getEstablishmentRatings(
  pageNum: number
): Promise<EstablishmentsType> {
  return fetch(
    `http://api.ratings.food.gov.uk/Establishments/basic/${pageNum}/10`,
    { headers: { "x-api-version": "2" } }
  ).then((res) => res.json());
}

export function getEstablishmentsByAuthority(
  authority: string,
  pageNum: number
): Promise<EstablishmentsType> {
  return fetch(
    `http://api.ratings.food.gov.uk/Establishments?localAuthorityId=${authority}&pageSize=10&pageNumber=${pageNum}`,
    { headers: { "x-api-version": "2" } }
  ).then((res) => res.json())
}

export function getAuthorities(
  pageNum: number
): Promise<AuthoritiesType> {
  return fetch(
    `http://api.ratings.food.gov.uk/Authorities/basic/${pageNum}/10`,
    { headers: { "x-api-version": "2" } }
  ).then((res) => res.json())
}

export function getAuthoritiesByName(
  name: string,
): Promise<AuthoritiesType> {
  // neither pageSize or pageNumber parameters are being applied when sent to the API
  return fetch(
    `http://api.ratings.food.gov.uk/Authorities?name=${name}`,
    { headers: { "x-api-version": "2" } }
  ).then((res) => res.json())
}