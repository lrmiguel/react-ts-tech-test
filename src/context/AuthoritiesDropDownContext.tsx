import { createContext, useContext } from "react";

interface AuthoritiesDropdownContextProps {
    menuRef: React.MutableRefObject<HTMLDivElement | null>;
    hasMore: boolean | void;
    handleLoadMore: () => void;
    isFetchingNextPage: boolean;
}

export const AuthoritiesDropdownContext = createContext<AuthoritiesDropdownContextProps | undefined>(undefined);

export const useAuthoritiesDropdown = () => {
    const ctx = useContext(AuthoritiesDropdownContext);
    if (!ctx) throw new Error("useAuthoritiesDropdown must be used within AuthoritiesDropdownContext.Provider");
    return ctx;
  };