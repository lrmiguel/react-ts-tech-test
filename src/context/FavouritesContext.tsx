import React, { createContext, useContext, useState } from "react";

export interface Favourite {
  id: string;
  name: string;
  rating: string;
}

interface FavouritesContextType {
  favourites: Favourite[];
  toggleFavourite: (fav: Favourite) => void;
  removeFavourite: (id: string) => void;
  isFavourited: (id: string) => boolean;
}

const FavouritesContext = createContext<FavouritesContextType | undefined>(undefined);

export const FavouritesProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [favourites, setFavourites] = useState<Favourite[]>([]);

  const toggleFavourite = (fav: Favourite) => {
    setFavourites((prev) =>
      prev.some((f) => f.id === fav.id)
        ? prev.filter((f) => f.id !== fav.id)
        : [...prev, fav]
    );
  };

  const removeFavourite = (id: string) => {
    setFavourites((prev) => prev.filter((f) => f.id !== id));
  };

  const isFavourited = (id: string) => {
    return favourites.some((f) => f.id === id);
  };
  return (
    <FavouritesContext.Provider value={{ favourites, toggleFavourite, removeFavourite, isFavourited }}>
      {children}
    </FavouritesContext.Provider>
  );
};

export const useFavourites = () => {
  const ctx = useContext(FavouritesContext);
  if (!ctx) throw new Error("useFavourites must be used within FavouritesProvider");
  return ctx;
};
