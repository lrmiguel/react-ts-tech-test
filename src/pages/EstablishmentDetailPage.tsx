import { useState } from "react";
import { useParams, useHistory } from "react-router-dom";
import { EstablishmentDetailType, getEstablishmentById } from "../api/ratingsAPI";
import { useQuery } from "react-query";
import styles from "./EstablishmentDetailPage.module.css";

const EstablishmentDetailPage = () => {
    const [error, setError] =
        useState<{ message: string; [key: string]: string }>();
    const { id } = useParams<{ id: string }>();
    const history = useHistory();

    const { data, isLoading } = useQuery<EstablishmentDetailType, Error>(
        ["establishment", id],
        async () => {
            try {
                return await getEstablishmentById(id);
            } catch (error) {
                throw new Error("Failed to fetch establishment details. Please try again later.");
            }
        },
        {
          keepPreviousData: true,
          refetchOnWindowFocus: false,
          staleTime: 5 * 60 * 1000,
          cacheTime: 10 * 60 * 1000,
          onError: (error) => {
            setError({ message: error.message });
          }
        }
    );
    const establishment = data;

    if (isLoading) {
        return <div className={`${styles.layoutStyle} ${styles.loading}`}>
            <strong>Loading...</strong>
        </div>;
    }
    if (error) {
        return <div className={styles.layoutStyle}>Error: {error?.message}</div>; 
    }

    const formatUKAddress = (establishment: EstablishmentDetailType) => {
        const lines = [
          establishment.AddressLine1,
          establishment.AddressLine2,
          establishment.AddressLine3,
          establishment.AddressLine4,
        ].filter(Boolean);
      
        return (lines.length > 0)? lines.join(", ") : "N/A";
    }

    const formatDate = (date: string) => {
        const utcDate = new Date(date);
        return utcDate.toLocaleDateString("en-GB");
    }

    return (!isLoading &&
        (
            <div className={styles.layoutStyle}>
                {!!error && <div>{error}</div>}
                <h2 style={{ textAlign: "center", marginBottom: "5px" }}>Establishment Details</h2>
                <div style={{ flexGrow: 1 }}>
                    <h3>{establishment?.BusinessName}</h3>
                    <p><strong>Address:</strong> {formatUKAddress(establishment!)}</p>
                    <p><strong>Rating:</strong> {establishment?.RatingValue}</p>
                    <p><strong>Date of Inspection:</strong> {formatDate(establishment!.RatingDate)}</p>
                </div>
                <div style={{textAlign: "center"}}>
                    <button className={styles.buttonStyle} onClick={() => history.push("/")}>Go back</button>
                </div>
            </div>
        )
    );
};
export default EstablishmentDetailPage;