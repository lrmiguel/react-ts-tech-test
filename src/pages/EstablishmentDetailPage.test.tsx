import { act, render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter, Route } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "react-query";
import * as ratingsAPI from "../api/ratingsAPI";
import EstablishmentDetailPage from "./EstablishmentDetailPage";

// Mock the ratingsAPI module
jest.mock("../api/ratingsAPI", () => ({
    getEstablishmentById: jest.fn(),
}));
import userEvent from "@testing-library/user-event";
import { FavouritesProvider } from "../context/FavouritesContext";

const createQueryClient = () => {
    return new QueryClient({
        defaultOptions: {
          queries: {
            retry: false,
            cacheTime: 0, // disable cache
            staleTime: 0,
          },
        },
      });
}
const renderWithProviders = (id: string, queryClient: QueryClient) => {
    return render(
        <MemoryRouter initialEntries={[`/establishments/${id}`]}>
            <Route path="/establishments/:id">
                <QueryClientProvider client={queryClient}>
                    <FavouritesProvider>
                        <EstablishmentDetailPage />
                    </FavouritesProvider>
                </QueryClientProvider>
            </Route>
        </MemoryRouter>
    );
};

describe("EstablishmentDetailPage", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });
    
    it("matches snapshot", async () => {
        const mockData = {
            BusinessName: "Test Business",
            AddressLine1: "123 Test Street", 
            AddressLine2: "Test Area",
            AddressLine3: "",
            AddressLine4: "",
            RatingValue: "5",
            RatingDate: "2023-01-01T00:00:00",
        };
        (ratingsAPI.getEstablishmentById as jest.Mock).mockResolvedValue(mockData);

        const { container } = renderWithProviders("123", createQueryClient());

        await waitFor(() => {
            expect(screen.getByText("Test Business")).toBeInTheDocument();
        });

        expect(container).toMatchSnapshot();
    });

    it("renders loading state initially", () => {
        (ratingsAPI.getEstablishmentById as jest.Mock).mockReturnValue(new Promise(() => {}));
        renderWithProviders("123", createQueryClient());

        expect(screen.getByText("Loading...")).toBeInTheDocument();
    });

    it("renders establishment details when data is fetched successfully", async () => {
        const mockData = {
            BusinessName: "Test Business",
            AddressLine1: "123 Test Street",
            AddressLine2: "Test Area",
            AddressLine3: "",
            AddressLine4: "",
            RatingValue: "5",
            RatingDate: "2023-01-01T00:00:00",
        };
        (ratingsAPI.getEstablishmentById as jest.Mock).mockImplementation(() => {
            return new Promise((resolve) =>
                setTimeout(() => resolve(mockData), 100)
            );
        });

        renderWithProviders("123", createQueryClient());

        await act(() => new Promise((resolve) => setTimeout(resolve, 100)));
        await waitFor(() => {
            expect(screen.getByText("Establishment Details")).toBeInTheDocument();
            expect(screen.getByText("Test Business")).toBeInTheDocument();
            expect(screen.getByText(/123 Test Street.*Test Area/)).toBeInTheDocument();
            expect(screen.getByText("5")).toBeInTheDocument();
            expect(screen.getByText("01/01/2023")).toBeInTheDocument();
        });
    });

    it("renders error message when API call fails", async () => {
        (ratingsAPI.getEstablishmentById as jest.Mock).mockRejectedValue("Failed to fetch");

        renderWithProviders("123", createQueryClient());

        await waitFor(() => {
            expect(screen.getByText(/Error:/)).toBeInTheDocument();
        });
    });

    it("navigates back to the home page when 'Go back' button is clicked", async () => {
        const mockData = {
            BusinessName: "Test Business",
            AddressLine1: "123 Test Street",
            AddressLine2: "Test Area",
            AddressLine3: "",
            AddressLine4: "",
            RatingValue: "5",
            RatingDate: "2023-01-01",
        };
        (ratingsAPI.getEstablishmentById as jest.Mock).mockResolvedValue(mockData);

        renderWithProviders("123", createQueryClient());

        await waitFor(() => {
            expect(screen.getByText("Go back")).toBeInTheDocument();
        });

        userEvent.click(screen.getByText("Go back"));

        await waitFor(() => {
            expect(window.location.pathname).toBe("/");
        });
    });
    
});