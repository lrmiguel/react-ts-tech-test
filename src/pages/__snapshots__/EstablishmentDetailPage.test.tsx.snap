// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstablishmentDetailPage matches snapshot 1`] = `
<div>
  <div
    class="layoutStyle"
  >
    <h2
      style="text-align: center; margin-bottom: 5px;"
    >
      Establishment Details
    </h2>
    <div
      style="flex-grow: 1;"
    >
      <h3>
        Test Business
      </h3>
      <p>
        <strong>
          Address:
        </strong>
         
        123 Test Street, Test Area
      </p>
      <p>
        <strong>
          Rating:
        </strong>
         
        5
      </p>
      <p>
        <strong>
          Date of Inspection:
        </strong>
         
        01/01/2023
      </p>
    </div>
    <div
      style="text-align: center;"
    >
      <button
        class="buttonStyle"
      >
        Go back
      </button>
    </div>
  </div>
</div>
`;
